packages/@react-aria/toast/src/useToastRegion.ts

```diff
@@ -0,0 +1,79 @@
+import {AriaLabelingProps, DOMAttributes} from '@react-types/shared';
+import {focusWithoutScrolling, mergeProps} from '@react-aria/utils';
+import {getInteractionModality, useFocusWithin, useHover} from '@react-aria/interactions';
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import {RefObject, useEffect, useRef} from 'react';
+import {ToastState} from '@react-stately/toast';
+import {useLandmark} from '@react-aria/landmark';
+import {useLocalizedStringFormatter} from '@react-aria/i18n';
+
+export interface AriaToastRegionProps extends AriaLabelingProps {
+  /**
+   * An accessibility label for the toast region.
+   * @default "Notifications"
+   */
+  'aria-label'?: string
+}
+
+export interface ToastRegionAria {
+  /** Props for the landmark region element. */
+  regionProps: DOMAttributes
+}
+
+/**
+ * Provides the behavior and accessibility implementation for a toast region containing one or more toasts.
+ * Toasts display brief, temporary notifications of actions, errors, or other events in an application.
+ */
+export function useToastRegion<T>(props: AriaToastRegionProps, state: ToastState<T>, ref: RefObject<HTMLElement>): ToastRegionAria {
+  let stringFormatter = useLocalizedStringFormatter(intlMessages);
+  let {landmarkProps} = useLandmark({
+    role: 'region',
+    'aria-label': props['aria-label'] || stringFormatter.format('notifications')
+  }, ref);
+
+  let {hoverProps} = useHover({
+    onHoverStart: state.pauseAll,
+    onHoverEnd: state.resumeAll
+  });
+
+  let lastFocused = useRef(null);
+  let {focusWithinProps} = useFocusWithin({
+    onFocusWithin: (e) => {
+      state.pauseAll();
+      lastFocused.current = e.relatedTarget;
+    },
+    onBlurWithin: () => {
+      state.resumeAll();
+      lastFocused.current = null;
+    }
+  });
+
+  // When the region unmounts, restore focus to the last element that had focus
+  // before the user moved focus into the region.
+  // TODO: handle when the element has unmounted like FocusScope does?
+  // eslint-disable-next-line arrow-body-style
+  useEffect(() => {
+    return () => {
+      if (lastFocused.current && document.body.contains(lastFocused.current)) {
+        if (getInteractionModality() === 'pointer') {
+          focusWithoutScrolling(lastFocused.current);
+        } else {
+          lastFocused.current.focus();
```

Comment:
Yeah, if you have a toast visible and then close it, we don't want to scroll the page when using a mouse, but if you use a keyboard (e.g. focus the toast, then use page up/page down or arrow keys to scroll or something) and then close it we should probably show where focus went?

Reply:
Right, I'm just a bit torn about whether we should show where focus went immediately (aka current behavior) or to restore focus without scrolling since the user probably scrolled the page intentionally (focus scrolling could then happen when they perform a keyboard interaction afterwards). I'm fine with the current behavior, this is a pretty minor case tbh